### PR TITLE
Add /system/settings endpoint (SQPIT-1141)

### PR DESCRIPTION
**Pull request by supersven**
Currently, only `brig.options.settings.setRestrictUserCreation` is exposed.

This is a stripped down version of https://github.com/wireapp/wire-server/pull/2890 - That only exposes the setting that's known to be safe.

It will be used by the new Android app to decide if an account creation flow should be shown or not.

### Swagger

image

### Ticket

[SQPIT-1141](https://wearezeta.atlassian.net/browse/SQPIT-1141)

## Checklist

 - [X] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [X] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api/src/Wire/API/SystemSettings.hs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import qualified Data.Swagger as S
+import Imports hiding (head)
+import Servant.Swagger.Internal.Orphans ()
+import Test.QuickCheck
+import Wire.Arbitrary
+
+-- | Subset of `Brig.Options.Settings` that is safe to be shown in public.
+--
+-- Used to expose settings via the @/system/settings@ endpoint. Please check
+-- with security if you want to add settings here.
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/SystemSettings.hs:15]:
```suggestion
-- Used to expose settings via the @/system/settings@ endpoint. ALWAYS CHECK
-- WITH SECURITY IF YOU WANT TO ADD SETTINGS HERE.
```

File: services/brig/test/integration/API/SystemSettings.hs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    ]
+
+testGetSettings :: Opts -> Brig -> Http ()
+testGetSettings opts brig = liftIO $ do
+  expectResultForSetting Nothing False
+  expectResultForSetting (Just False) False
+  expectResultForSetting (Just True) True
+  where
+    expectResultForSetting :: Maybe Bool -> Bool -> IO ()
+    expectResultForSetting s expectedRes = do
```
Comment (fisx) [inline services/brig/test/integration/API/SystemSettings.hs:28]: Please avoid single-byte identifiers in the future!  (I know, it's a very short function, and I'm not doing this consistently either, but still...)

File: services/brig/test/integration/API/SystemSettings.hs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      let newOpts = opts & (optionSettings . restrictUserCreation) .~ s
+      queriedSettings <- withSettingsOverrides newOpts $ getSystemSettings brig
+      liftIO $
+        systemSettingsSetRestrictUserCreation queriedSettings @?= expectedRes
```
Comment (fisx) [inline services/brig/test/integration/API/SystemSettings.hs:32]:
```suggestion
        queriedSettings @?= SystemSettings expectedRes
```

It's shorter, and it fails for future extensions of `SystemSettings`, which is what we want.

**Review by fisx — APPROVED** [review]

**Comment by fisx** [comment]
ah, one more thing: can we make the end-point `/system/settings/unauthed`, so we can add `/system/settings/authed` for more sensitive settings later?  it's not a big barrier for the cloud instance, but for on-prem overcoming it usually requires access to employee credentials.

feel free to change the naming to something less silly.

**Review by fisx — CHANGES_REQUESTED** [review]
let's find out about how this settings overrides thing works first!

File: services/brig/test/integration/API/SystemSettings.hs (hunk 2 above)
Comment (supersven) [reply services/brig/test/integration/API/SystemSettings.hs]: Agreed :+1:

**Comment by supersven** [comment]
> ah, one more thing: can we make the end-point /system/settings/unauthed, so we can add /system/settings/authed for more sensitive settings later? it's not a big barrier for the cloud instance, but for on-prem overcoming it usually requires access to employee credentials.
>
> feel free to change the naming to something less silly.

@fisx I've changed it to `<version>/system/settings/unauthorized`. (As it's meant to be used by machines / apps a longer URL should IMHO be fine.) I'm not very creative... :laughing:

File: services/brig/test/integration/API/SystemSettings.hs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+testGetSettings :: Opts -> Brig -> Http ()
+testGetSettings opts brig = liftIO $ do
+  expectResultForSetting Nothing False
+  expectResultForSetting (Just False) False
+  expectResultForSetting (Just True) True
+  where
+    expectResultForSetting :: Maybe Bool -> Bool -> IO ()
+    expectResultForSetting s expectedRes = do
+      let newOpts = opts & (optionSettings . restrictUserCreation) .~ s
+      queriedSettings <- withSettingsOverrides newOpts $ getSystemSettings brig
```
[the remark this answers is not in the document]
Comment (supersven) [reply services/brig/test/integration/API/SystemSettings.hs]: The changes of the `brig` `Request` modifier function was mostly ignored. The only modification that counted was the addition of a version prefix.

E.g. port and host modifications were ignored. If you're really into it, the `MonadHttp` instances of `Session` (from `Network.Wai.Test.Internal`) and `SessionT` (`Bilge.TestSession`) should explain it: https://hoogle.zinfra.io/file/nix/store/9axalivlwwgzm4qmv7nd852hv6qm4jgm-bilge-0.22.0-doc/share/doc/bilge-0.22.0/html/Bilge-IO.html#t:MonadHttp

[the remark this answers is not in the document]
Comment (supersven) [reply services/brig/test/integration/API/SystemSettings.hs]: Ah, and thanks for bringing this up. The `brig` function was (almost) superfluous in this case. :+1:

**Comment by fisx** [comment]
> > ah, one more thing: can we make the end-point /system/settings/unauthed, so we can add /system/settings/authed for more sensitive settings later? it's not a big barrier for the cloud instance, but for on-prem overcoming it usually requires access to employee credentials.
> > feel free to change the naming to something less silly.
> 
> @fisx I've changed it to `<version>/system/settings/unauthorized`. (As it's meant to be used by machines / apps a longer URL should IMHO be fine.) I'm not very creative... laughing

creativity is not always a virtue.  `factchecker` is a better name for a fact checker app than `hyperfloyd`.  :)

File: services/brig/test/integration/API/SystemSettings.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/test/integration/API/SystemSettings.hs]: maybe you add a comment explaining this to the code, for future me?

**Review by fisx — APPROVED** [review]

File: services/brig/test/integration/API/SystemSettings.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (supersven) [reply services/brig/test/integration/API/SystemSettings.hs]: Well, the *future me*s with bad memory. I know those... got one, too :wink: 

(Comment added)

**Pull request merged**